File: tests/apple_street_preview.py
#!/usr/bin/env python3
"""Apple preview lifecycle tests using a stub SDK; no Apple quota is consumed."""
from pathlib import Path
from playwright.sync_api import sync_playwright
ROOT=Path(__file__).resolve().parent.parent
MODULE=(ROOT/'static/apple-street-preview.js').read_text()
FRAME=(ROOT/'static/apple-street-frame.html').read_text()
HTML='''<style>.tile{position:relative;width:300px;height:150px;margin:10px}.apple-street-preview{position:absolute;inset:0}[hidden]{display:none!important}</style>
<div id="tiles"></div><script>window.testCreated=0;window.testDestroyed=0;window.testLoading=0;window.testMaxLoading=0;</script><script src="/apple.js"></script>'''
SDK='''
class Preview extends EventTarget {
 constructor(host,coordinate,options){super();parent.testCreated++;parent.testLoading++;
 parent.testMaxLoading=Math.max(parent.testMaxLoading,parent.testLoading);
 this.element=document.createElement('div');this.element.textContent='Apple SDK test fixture';host.appendChild(this.element);
 this.openDialog=false;this.readyState='loading';this.pending=true;
 parent.testCoordinates?.push(coordinate.latitude);
 parent.testOptions=options;
 this.element.addEventListener('click',()=>{this.openDialog=true;this.dispatchEvent(new Event('enter-dialog'));});
 this.timer=setTimeout(()=>{this.pending=false;parent.testLoading--;this.readyState=coordinate.latitude<39?'error':'complete';this.dispatchEvent(new Event(this.readyState==='error'?'error':'load'));},coordinate.latitude===41?1000:30);}
 destroy(){clearTimeout(this.timer);if(this.pending){parent.testLoading--;this.pending=false;}parent.testDestroyed++;this.element.remove();this.readyState='destroyed';}
}
window.mapkit={LookAroundPreview:Preview,Coordinate:class{constructor(lat,lng){this.latitude=lat;this.longitude=lng;}}};
window[document.currentScript.dataset.callback]();'''
def main():
 with sync_playwright() as p:
  for name in ('chromium','webkit'):
   browser=getattr(p,name).launch()
   page=browser.new_page(viewport={'width':390,'height':844})
   errors=[];sdk_requests=[]
   page.on('pageerror',lambda e:errors.append(str(e)))
   def app(route):
    route.fulfill(content_type='application/javascript' if '/apple.js' in route.request.url else 'text/html',body=MODULE if '/apple.js' in route.request.url else FRAME if '/static/apple-street-frame.html' in route.request.url else HTML)
   page.route('http://localhost/**',app)
   def sdk(route):
    sdk_requests.append(True);route.fulfill(content_type='application/javascript',body=SDK)
   page.route('https://cdn.apple-mapkit.com/**',sdk)
   page.goto('http://localhost/?photos=apple')
   def render(lat=40,count=1):
    page.evaluate('''([lat,count])=>{const host=document.getElementById('tiles');host.innerHTML=Array.from({length:count},()=>'<div class="tile">'+AppleStreetPreview.html({lat,lng:-73,a:'Test address'})+'</div>').join('');AppleStreetPreview.watch(host);}''',[lat,count])
   render()
   page.get_by_text('Connect Apple Maps to preview photos',exact=True).wait_for()
   assert not sdk_requests
   page.evaluate("sessionStorage.setItem('fac.apple_maps_token','test.web.token')")
   page.reload();render(count=4)
   page.wait_for_function('parent.testCreated===4 && parent.testLoading===0')
   assert len(sdk_requests)==4
   # 5 since 2026-09-15: three at a time left visible cards queued behind each
   # other, and Safari never finishes an off-screen preview, so the slot that
   # used to warm the next row jammed the queue for 45 s.
   assert page.evaluate('parent.testMaxLoading')<=5
   assert page.locator('.apple-street-status:not([hidden])').count()==0
   assert page.evaluate('parent.testOptions.isNavigationEnabled && parent.testOptions.isScrollEnabled && parent.testOptions.isZoomEnabled')
   attrs=page.locator('iframe').first.content_frame.locator('script[data-libraries]').evaluate('(el)=>({libraries:el.dataset.libraries,token:el.dataset.token})')
   assert attrs=={'libraries':'look-around','token':'test.web.token'}
   first_frame=page.locator('iframe').first
   first_frame.content_frame.get_by_text('Apple SDK test fixture').click()
   page.wait_for_function('document.querySelector("iframe").matches(":popover-open")')
   assert first_frame.evaluate('(f)=>f.clientWidth')==390
   page.evaluate('AppleStreetPreview.close(document.getElementById("tiles"))')
   page.wait_for_function('!document.querySelector("iframe").hasAttribute("popover")')
   assert page.locator('iframe').count()==4, 'Closing the viewer should keep its iframe available for deferred cleanup'
   first_frame.content_frame.get_by_text('Apple SDK test fixture').click()
   page.wait_for_function('document.querySelector("iframe").matches(":popover-open")')
   page.keyboard.press('Escape')
   page.wait_for_function('!document.querySelector("iframe").hasAttribute("popover")')
   assert first_frame.evaluate('(f)=>f.clientWidth')==300
   render(38)
   page.get_by_text('Apple street imagery is unavailable here',exact=True).wait_for()
   assert page.locator('iframe').count()==0
   page.evaluate("document.getElementById('tiles').style.marginTop='2000px'")
   before=page.evaluate('parent.testCreated');render(40)
   page.wait_for_timeout(100)
   assert page.evaluate('parent.testCreated')==before
   page.locator('.tile').scroll_into_view_if_needed()
   page.wait_for_function('parent.testLoading===0 && document.querySelector(".apple-street-status").hidden')
   assert page.evaluate('parent.testCreated')==before+1
   # Warm the next row inside the actual scrolling grid; do not warm distant rows.
   page.reload()
   page.evaluate("""()=>{window.testCoordinates=[];const host=document.getElementById('tiles');
    host.innerHTML='<div id="grid" style="height:170px;overflow:auto"></div>';
    const grid=document.getElementById('grid');grid.innerHTML=Array.from({length:10},(_,i)=>'<div class="tile">'+AppleStreetPreview.html({lat:40+i/100,lng:-73})+'</div>').join('');AppleStreetPreview.watch(grid);}""")
   page.wait_for_function('window.testCoordinates.length>=3 && window.testLoading===0')
   assert page.evaluate('window.testCoordinates[0]')==40
   assert page.evaluate('window.testCoordinates.length')<=4
   assert page.evaluate('window.testMaxLoading')<=3
   warmed=page.evaluate('window.testCreated')
   page.evaluate("document.getElementById('grid').scrollTop=160")
   page.locator('#grid .tile').nth(1).locator('.apple-street-status[hidden]').wait_for(state='attached')
   # A slow photo left far behind is cancelled so it cannot block a new row.
   page.reload()
   page.evaluate("""()=>{window.testCoordinates=[];const host=document.getElementById('tiles');
    host.innerHTML='<div id="grid" style="height:150px;overflow:auto"></div>';
    const grid=document.getElementById('grid');grid.innerHTML=Array.from({length:14},(_,i)=>'<div class="tile">'+AppleStreetPreview.html({lat:i===0?41:42+i/100,lng:-73})+'</div>').join('');AppleStreetPreview.watch(grid);}""")
   page.wait_for_function('window.testCoordinates.includes(41)')
   page.evaluate("document.getElementById('grid').scrollTop=1600")
   page.wait_for_function("!document.querySelector('[data-apple-lat=\"41\"] iframe')")
   page.wait_for_function('window.testCoordinates.some(lat=>lat>=42.09)')
   before_requests=len(sdk_requests)
   # Default local and production pages use Apple; Mapillary stays opt-in locally.
   page.goto('http://localhost/')
   assert page.evaluate('AppleStreetPreview.enabled')
   page.goto('http://localhost/?photos=mapillary')
   assert not page.evaluate('AppleStreetPreview.enabled')
   page.route('https://findacrib.com/**',app)
   page.goto('https://findacrib.com/')
   assert page.evaluate('AppleStreetPreview.enabled')
   render()
   page.get_by_text('Street photo temporarily unavailable',exact=True).wait_for()
   assert not page.locator('a[href="/apple-preview/"]').count()
   assert len(sdk_requests)==before_requests
   page.add_init_script("window.APPLE_MAPS_TOKEN='test.web.token'")
   page.reload();render()
   page.wait_for_function('document.querySelector(".apple-street-status").hidden')
   assert page.locator('iframe').count()==1
   page.evaluate('AppleStreetPreview.release(document.getElementById("tiles"))')
   assert page.locator('iframe').count()==0
   page.route('https://unrelated.example/**',app)
   page.goto('https://unrelated.example/')
   assert not page.evaluate('AppleStreetPreview.enabled')
   assert not errors, errors
   print(f'PASS {name}: token setup, isolated frames, lazy initialization, concurrency, cleanup, coverage errors, hostname gating',flush=True)
   browser.close()
if __name__=='__main__':main()

File: static/apple-street-preview.js
/* Apple Look Around photos. Uses Apple's supported renderer;
   does not extract, store, or rehost Apple imagery. */
(() => {
  const local = ['localhost', '127.0.0.1', '[::1]'].includes(location.hostname);
  const enabled = (local || ['findacrib.com', 'www.findacrib.com'].includes(location.hostname))
    && !(local && new URLSearchParams(location.search).get('photos') === 'mapillary');
  let storedToken = '';
  if (local) { try { storedToken = sessionStorage.getItem('fac.apple_maps_token') || ''; } catch (_) {} }
  const token = (window.APPLE_MAPS_TOKEN || storedToken).trim();
  const records = new Map(), queue = [];
  let loading = 0;
  const observers = new Map();
  const esc = s => String(s ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
  function destroy(record) {
    record.cancel?.();
    try { record.view?.destroy(); } catch (_) {}
    record.view = null;
  }
  function prune() {
    records.forEach((record, el) => {
      if (!el.isConnected) { record.observer?.unobserve(el); destroy(record); records.delete(el); }
    });
    // Retain a few offscreen previews so small scrolls don't reload them.
    const mounted = [...records.values()].filter(r => r.view);
    for (const record of mounted) {
      if (mounted.filter(r => r.view).length < 12) break;
      if (!record.visible && !record.loading && !record.view.openDialog) destroy(record);
    }
  }
  async function mount(record) {
    const {el} = record, status = el.querySelector('.apple-street-status');
    if (!token) {
      status.innerHTML = local ? '<a href="/apple-preview/">Connect Apple Maps to preview photos</a>' : 'Street photo temporarily unavailable';
      record.failed = true; return;
    }
    status.hidden = false; status.textContent = 'Loading Apple street view…';
    record.loading = true;
    try {
      if (!el.isConnected || !record.visible) return;
      // MapKit supports one Look Around instance per document. Each tile gets
      // its own document and the browser caches the shared SDK files.
      const frame = document.createElement('iframe');
      frame.title = el.querySelector('.apple-street-canvas').getAttribute('aria-label');
      frame.src = '/static/apple-street-frame.html?v=3';
      frame.style.cssText = 'width:100%;height:100%;border:0;display:block;margin:0;padding:0';
      let expanded = false;
      function collapse() {
        if (!expanded) return;
        expanded = false;
        frame.hidePopover();
        frame.removeAttribute('popover');
        frame.style.width = frame.style.height = '100%';
        frame.contentWindow?.postMessage({type:'fac-apple-close'}, location.origin);
      }
      const escape = event => { if (event.key === 'Escape') collapse(); };
      record.view = {
        get openDialog() { return expanded; },
        close() { collapse(); },
        destroy() { collapse(); window.removeEventListener('keydown', escape); frame.remove(); }
      };
      window.addEventListener('keydown', escape);
      await new Promise(resolve => {
        let settled = false;
        const finish = message => {
          if (settled) return;
          settled = true; clearTimeout(timer); record.cancel = null;
          status.hidden = !message;
          if (message) { status.textContent = message; record.failed = true; }
          resolve();
        };
        const receive = event => {
          if (event.origin !== location.origin || event.source !== frame.contentWindow) return;
          if (event.data?.type !== 'fac-apple-state') return;
          const state = event.data.state;
          if (state === 'complete') finish('');
          if (state === 'error') finish('Apple street imagery is unavailable here');
          if (state === 'browser-error') finish('Street photos are unavailable in this browser');
          if (state === 'open' && !expanded && frame.showPopover) {
            expanded = true;
            frame.popover = 'manual';
            frame.style.width = '100vw'; frame.style.height = '100dvh';
            frame.showPopover();
          }
          if (state === 'close') collapse();
        };
        window.addEventListener('message', receive);
        const remove = record.view.destroy;
        record.view.destroy = () => { window.removeEventListener('message', receive); remove(); };
        const timer = setTimeout(() => finish('Apple street view took too long to load'), 45000);
        record.cancel = () => finish('');
        frame.addEventListener('load', () => frame.contentWindow.postMessage({
          type:'fac-apple-init', token,
          lat:Number(el.dataset.appleLat), lng:Number(el.dataset.appleLng)
        }, location.origin), {once:true});
        el.querySelector('.apple-street-canvas').appendChild(frame);
      });
      if (record.failed) destroy(record);
    } catch (error) {
      status.textContent = error.message || 'Apple street view is unavailable';
      status.hidden = false; record.failed = true; destroy(record);
    } finally { record.loading = false; }
  }
  function onScreen(record) {
    const rect = record.el.getBoundingClientRect();
    const bounds = record.root?.getBoundingClientRect() || {top:0, bottom:innerHeight, left:0, right:innerWidth};
    return rect.width > 0 && rect.height > 0 && rect.bottom > Math.max(0, bounds.top)
      && rect.top < Math.min(innerHeight, bounds.bottom) && rect.right > Math.max(0, bounds.left)
      && rect.left < Math.min(innerWidth, bounds.right);
  }
  function drain() {
    prune();
    // Load visible photos first. Two slots may warm the next row.
    queue.sort((a,b) => Number(onScreen(b)) - Number(onScreen(a)));
    while (loading < 5 && queue.length) {
      const record = queue[0];
      if (!record.el.isConnected || !record.visible || record.view || record.failed || record.loading) {
        queue.shift(); record.queued = false; continue;
      }
      const prefetch = !onScreen(record);
      // Sorted on-screen first, so if this one is off screen none behind it
      // are either, and a record that already stalled off screen waits for the
      // scroll rather than taking a slot again.
      if (prefetch && (record.needsVisible
          || [...records.values()].filter(r => r.loading && !onScreen(r)).length >= 2)) break;
      queue.shift(); record.queued = false;
      loading++;
      // Safari does not run an iframe that is off screen: a preview started
      // below the fold never reports back, held its slot until the 45 s
      // timeout, and every card behind it waited (2026-09-15 — three photos
      // loaded and the rest sat on "Loading…"). Hand the slot back after 6 s
      // and let that card mount for real when it scrolls into view.
      const guard = prefetch ? setTimeout(() => {
        if (!record.loading || onScreen(record)) return;
        record.needsVisible = true;
        const status = record.el.querySelector('.apple-street-status');
        destroy(record);
        if (status) { status.hidden = false; status.textContent = 'Loading Apple street view…'; }
      }, 6000) : null;
      mount(record).finally(() => {
        clearTimeout(guard);
        loading--;
        if (record.visible && record.el.isConnected) enqueue(record);
        drain();
      });
    }
  }
  function enqueue(record) {
    if (record.queued || record.loading || record.view || record.failed) return;
    record.queued = true; queue.push(record);
  }
  function observerFor(root) {
    if (!enabled || !('IntersectionObserver' in window)) return null;
    if (!observers.has(root)) {
      observers.set(root, new IntersectionObserver(entries => {
        entries.forEach(entry => {
          const record = records.get(entry.target);
          if (!record) return;
          record.visible = entry.isIntersecting;
          if (record.visible) enqueue(record);
          // A quick scroll must not leave the new row waiting on old requests.
          else if (record.loading && !record.view?.openDialog) destroy(record);
        });
        drain();
      }, {root, rootMargin:root ? '350px 0px' : '0px'}));
    }
    return observers.get(root);
  }
  function watch(root) {
    if (!enabled) return;
    prune();
    root.querySelectorAll('[data-apple-lat]').forEach(el => {
      if (records.has(el)) return;
      const scrollRoot = el.closest('#grid');
      const observer = observerFor(scrollRoot);
      const record = {el, root:scrollRoot, observer, visible:!observer, view:null, queued:false, loading:false, failed:false};
      records.set(el, record);
      // Apple's controls open its own viewer; the address/body still opens building details.
      el.addEventListener('click', e => e.stopPropagation());
      if (observer) observer.observe(el); else enqueue(record);
    });
    drain();
  }
  function html(rec) {
    if (!enabled) return '';
    const lat=Number(rec.lat), lng=Number(rec.lng);
    if (rec.lat == null || rec.lng == null || rec.lat === '' || rec.lng === ''
      || !Number.isFinite(lat) || !Number.isFinite(lng) || Math.abs(lat)>90 || Math.abs(lng)>180)
      return '<span class="apple-street-status">Street location unavailable</span>';
    return `<div class="apple-street-preview" data-apple-lat="${lat}" data-apple-lng="${lng}">
      <div class="apple-street-canvas" aria-label="Apple street preview near ${esc(rec.a || rec.address || rec.name || 'this property')}"></div>
      <span class="apple-street-status" role="status">Loading Apple street view…</span>
    </div>`;
  }
  function release(root) {
    records.forEach((record, el) => {
      if (!root.contains(el)) return;
      record.observer?.unobserve(el); destroy(record); records.delete(el);
    });
  }
  function close(root) {
    records.forEach((record, el) => {
      if (root.contains(el)) record.view?.close?.();
    });
  }
  window.AppleStreetPreview = {enabled, provider:'apple', html, watch, release, close};
})();
